### ISSUE TITLE
Implementing coclasses is weird.

### ISSUE THREAD

**Issue by bclothier** [WaynePhillipsEA/twinbasic#603]
**Describe the bug**
twinBasic allows one to implement a coclass. However, this is a case of VBx saying one thing but doing something different. As a consequence, implementing coclass may run into problems. Fundamentally, we work with interfaces and a coclass is usually nothing more than a list of all interfaces it may provide. Furthermore, a coclass can identify a default interface and also a default source interface. Therefore, it can be reasonably inferred that implementing a coclass should mean the same `default` and `source` attributes applied to the implemented coclass' interfaces ought to be also applied to the implementing coclass' interfaces. It is not and VBx has never worked that way.

Even if we addressed this gap, it still opens up a problem -- the rule only allows for at most one `default` interface and one `default, source` interface. But we can implement more than one coclasses in twinBasic. What does that actually mean?

**To Reproduce**
1. Using this IDL, compile into a TLB with the MIDL compiler:
```
[
  uuid(B8C22F55-F703-4610-ACC3-5D9B91B0BC30)
]
library FooLibrary
{
	importlib("stdole2.tlb");
	
	[
		uuid(ECED7F40-0354-47DB-A4E6-B1E65F2CD6D6),
		dual,
		oleautomation
	]
	interface IFoo : IDispatch
	{
		  HRESULT _stdcall Foo();
	}     
	
	[
		uuid(366FE734-02A8-4A76-BDE8-4CCAEB0311E9)
	]
	dispinterface IFooEvents
	{
		properties:
		methods:
		[id(1)]
		void AfterFoo();
	}
	
	[
	  uuid(E4072424-303A-49FA-88C8-CA6EA9A28181)
	]
	coclass Foo {
		[default] interface IFoo;
		[default, source] dispinterface IFooEvents;
	};
}
```
2. In an ActiveX project import the compiled TLB file and then generate a class module:
```
Public Class Bar
    Implements FooLibrary.Foo
    
    Public Sub Foo() Implements FooLibrary.Foo.Foo
    End Sub
End Class
```
3. Build the project.
4. Using a type library viewer tool (e.g. oleview or olewoo), observe that the generated type library for the built project has this entry:
```
coclass Bar {
    [default] interface _Bar;
};
```

**Expected behavior**
However, if implementing a coclass is to be allowed, it is more likely that there's an expectation that the implementing coclass should match the implemented coclass in all the forms, including the attributes applied to its interfaces. However, tB generates a new interface and makes it the default. I don't think that is right. Unfortunately, we are constrained by the VBx compatibility here. 

At the minimum, the implemented coclass' interface (in this case, `IFoo`) ought to be in the implementing coclass' interface list. 

We also need to consider whether the `default` attribute ought to be copied. Not sure what VBx will do here. 

**Additional context**
I'm aware that the source interfaces aren't supported yet but I'm including this here for the completeness' sake. 

That said, I find VBx's current behavior to be very confusing. From the statement: `Implements SomeClass`, I would expect the implementing class to mirror _all_ of its interface and the attributes applied to the interfaces. If I didn't want that, I'd have used `Implements ISomeClass` and (hypothetically) `Implements WithEvents ISomeClassEvents` instead. But the VBx's behavior seem to be to simply implement the interfaces but not the source interfaces which makes the statement `Implements SomeClass` a complete lie.

To remedy that confusing behavior and keeping backward compatibility, we might benefit having the compiler warning on the `Implements SomeClass` and alerting that it's actually just `Implements ISomeClass`. We can then make the decision to change the source code and be more explicit. That would also help in alerting that we do not have a full fidelity between our implementing coclass and the implementing coclass definition so we can then correct this by writing necessary statements or attributes. For instance,  we can specify the behavior with a new attribute:
```
[CoClassInherits(Default:True, Source:True, DefaultSource:True)]
Implements SomeClass
```

We can then leverage the CodeLen feature so that an existing `Implements SomeClass` will get decorated with `[CoClassInherits(Default: False, Source:False, DefaultSource:False)]` without actually changing the source code.

Alternatively, the warning should make us use the interfaces directly to get more control over how our coclass gets decorated:
```
[Default]
Implements ISomeClass

[Source]
Implements WithEvents ISomeClassEvents
```

This approach avoids the potential problems arising when implementing more than 1 coclasses which may have each their own `default`, preventing our class from completely mirroring the 2nd & beyond coclasses.

Regardless of how it is done, the ultimate goal here is to help avoid the confusion that comes around with using the `Implements` statement and the fact that VBx go too far in hiding details about interfaces that are generated and how it is handled by the coclasses.

**Desktop (please complete the following information):**
 - OS: Windows 10
 - twinBASIC compiler version 0.13.54

**Comment by EduardoVB** [comment]
It would be more than welcome to be able to have `Implements WidthEvents`, something not possible in VBx, and a limitation that needs workarounds.

**Comment by wqweto** [comment]
This proposal is pretty vague. Is the new `Implements WithEvents` syntax that is being proposed? New attributes on classes to control typelib generation?

**Comment by bclothier** [comment]
This originally started out as a bug report but I realized I wasn't 100% sure if that is a bug. This looks like a bug to me. From OP:

```
Implements Foo
```

generates the typelib:
```
coclass Bar {
    [default] interface _Bar;
};
```

It looks like the `interface IFoo` is missing from the `coclass Bar` even though it's supposed to implement the `Foo`. But I realized that I wasn't sure because that might be how VBx already behave currently. 

If that is how it behaves, then it feels like the `Implements` statement is a lie because it's saying one thing but doing something different. The backward compatibility then becomes a constraint here since if we change the behavior of the `Implements` statement, it would break the backward compatibility. After that, I don't have a specific recommendation, only a request that tB be more honest with what it's actually doing in response to `Implements Foo` statement.

**Comment by wqweto** [comment]
> But I realized that I wasn't sure because that might be how VBx already behave currently.

This is precisely what VB6 does.

> If that is how it behaves, then it feels like the Implements statement is a lie because it's saying one thing but doing something different.

This is the same way `Dim Obj As Foo` is a lie -- it should be `Dim Obj As IFoo` because the pointer can be to an interface not to a coclass.

Alas this is the BASIC approach in VBx to hide the difference between interface and a coclass, the latter being using interchangably with its default interface.

**Comment by mburns08109** [comment]
OK, I'm a little lost in the technical details, here.
Can someone give me a very concise definition of how a coclass is different from a plain olde class?
My understanding right now is that "it's the same, but different", and that doesn't help me grok this topic well.

**Comment by bclothier** [comment]
> Alas this is the BASIC approach in VBx to hide the difference between interface and a coclass, the latter being used interchangably with its default interface.

Coming from C# + COM interop, it does feels like it's a letdown because I now have to "know" that the language is hiding details from me and now I'm fighting with the language. If I say that a class implements this class, I expect the class to inherit everything about this class. If I only wanted the interface, I would have specified that the class implements the interface instead. 

But with this confounding behavior of pretending that a coclass is a thing we can work with when in actuality we're dealing with the coclass' default interface makes it harder to author a meaningful implementation. I can't use `Implements Foo` in its present form because of the fact that it has nothing to do with how my types will be exposed to other COM clients. I do not think `Implements WithEvents Foo` is an improvement in this context because it only says that it should raise the same events but says nothing about the type library generation. Critically, it'd need to preserve the behavior of `Implements` statement to avoid behaving drastically different when the `WithEvents` is added. 

To avoid this from becoming a landmine to trip up the unwary programmer who may be accustomed to interfaces or working with COM more directly, I think the existing `Implements` statement ought to have a CodeLen feature to call the attention to the fact that it's not doing what it is saying. We have to treat the `Implements` statement as meaningful only within the context of the language and not the broader context of COM types and look elsewhere for handling the type library generation (e.g. attributes, maybe). 

I'm inclined to think the reason `Implements` wasn't widely used (at least I've not seen widespread usage of it in the VBx world and especially not among VBA codebases) is precisely because of all the obfuscation going on with the types being implemented. I think tB need to address this if we are to have an excellent COM authoring solution, IMO.

**Comment by bclothier** [comment]
> Can someone give me a very concise definition of how a coclass is different from a plain olde class?

We are talking about 2 different levels here. A VBx class is a construct that exists within the VBx language. 

But every VBx class is also a COM coclass. This is not exposed to you at the VBx level but would be if you were working at the COM level.

A COM coclass is nothing more than a contract that essentially says that if you have an instance of that coclass, you can use any COM interfaces provided by that coclass. Most of time, you'll use the instance through its default interface. 

But VBx hides that complexity from you, so your code `Dim app As Excel.Application: Set app = New Excel.Application` is actually creating an instance of `Excel.Application` coclass and then providing you with the interface `Excel.[_Application]` (which is the default interface of the coclass). Thus, when you do `app.Visible = True`, you are actually calling a method on the interface... `Excel.[_Application].Visible`.

That is why implementing a coclass is weird. It's more proper to think of it as a list of interfaces that it can provide, and only interfaces can be actually implemented. So if I'm implementing a class, I expect to be implementing _all_ of its interface.

In fact, that is what tB doing (except for source interfaces which are for event handling). So if you implement a class in tB, you will be required to implement all of the members from all of the interfaces that the class has. But your custom class will not expose the same set of interfaces at the COM level and that's where the issue lie.

**Comment by mansellan** [comment]
I don't really know much at the COM level, but is this something that could be addressed with attributes somehow? Just an idea.

**Comment by bclothier** [comment]
Yes, attributes would work but that doesn't address what I think is more fundamental problem --- the code `Implements` not doing what it says on the tin. That's why I'm partial to the idea of using CodeLen in conjunction with an attribute to help and make it more explicit.

**Comment by mburns08109** [comment]
OK, perhaps I'm not recalling the details here again, but how does VBx presently allow a class to have multiple interfaces that are NOT inherited via an Implements - or can't it?

I mean:
=== Myclass ===
--MyInterfaceX--
Method A
Method B
--MyInterfaceY--
MethodX
MethodY
Implements MyOtherClass as MOC 'actually is Myotherclass._Default
MOC.MethodN
MOC.MethodM
===End of MyClass===

So MyClass has 6 defined methods, but only some are accessible depending on the interface that client code implements...?

**Comment by EduardoVB** [comment]
Of course, the client access the object from one interface, and that interface has the methods that it has and no other ones.
Or you can set the same object to different variables, and access the same object from different interfaces, each one with its methods.

Example:

```
    Dim p As StdPicture
    Dim p2 As IPicture
    
    Set p = LoadPicture("path.jpg")
    Set p2 = p
    
    Debug.Print p.Height, p2.Attributes
```

**Comment by mburns08109** [comment]
> OK, perhaps I'm not recalling the details here again, but how does VBx presently allow a class to have multiple interfaces that are NOT inherited via an Implements - or can't it?
> 
> I mean: === Myclass === --MyInterfaceX-- Method A Method B --MyInterfaceY-- MethodX MethodY Implements MyOtherClass as MOC 'actually is Myotherclass._Default MOC.MethodN MOC.MethodM ===End of MyClass===
> 
> So MyClass has 6 defined methods, but only some are accessible depending on the interface that client code implements...?

OK, so what I meant there was: HOW does VBx CREATE the multiple instances per class NOW (WITHOUT using Implements to enable an interface defined externally to the class) - i.e. with what syntax at present?

**Comment by EduardoVB** [comment]
> OK, so what I meant there was: HOW does VBx CREATE the multiple interfaces per class NOW (WITHOUT using Implements to enable an interface defined externally to the class) - i.e. with what syntax at present?

The only way IFAIK is via Implements.

**Comment by bclothier** [comment]
> OK, so what I meant there was: HOW does VBx CREATE the multiple interfaces per class NOW (WITHOUT using Implements to enable an interface defined externally to the class) - i.e. with what syntax at present?

See, that's the confusing thing. Because there is no language construct that refers to an interface, you would just create a class module without any implementation. Maybe [this article will help](https://rubberduckvba.wordpress.com/2020/02/27/vba-classes-gateway-to-solid/).

**Comment by mburns08109** [comment]
> > OK, so what I meant there was: HOW does VBx CREATE the multiple interfaces per class NOW (WITHOUT using Implements to enable an interface defined externally to the class) - i.e. with what syntax at present?
> 
> See, that's the confusing thing. Because there is no language construct that refers to an interface, you would just create a class module without any implementation. Maybe [this article will help](https://rubberduckvba.wordpress.com/2020/02/27/vba-classes-gateway-to-solid/).

OK, but that's exactly what I was getting at.
We have no "interface-level" language support in VBx presently, right?
But we do have it in vb.net now: https://docs.microsoft.com/en-us/dotnet/visual-basic/programming-guide/language-features/interfaces/ 

So, why don't we add this syntax to tB as our primary means of addressing these coClass issues?

**Comment by bclothier** [comment]
That is already the case in tB. You can create an interface explicitly in tB.

```
Public Interface IFoo
  Sub Foo()
End Interface 
```

The issue here is not whether we can do that in tB (we can), but rather whether it’s reflected so in tne generated type library at the COM level. Apparently becauseq VBx didn’t, that requires considerations to give the necessary control over the shape of the generated type library.

**Comment by mburns08109** [comment]
Hmm...and might one of the implications there mean that we may also need to consider adding the MODULE syntax from vb.net as well?
https://docs.microsoft.com/en-us/dotnet/visual-basic/language-reference/statements/module-statement

**Comment by mburns08109** [comment]
> That is already the case in tB. You can create an interface explicitly in tB.
> 
> ```
> Public Interface IFoo
>   Sub Foo()
> End Interface 
> ```
> 
> The issue here is not whether we can do that in tB (we can), but rather whether it’s reflected so in tne generated type library at the COM level. Apparently becauseq VBx didn’t, that requires considerations to give the necessary control over the shape of the generated type library.

Seems like sort of a no-brainer to me.
If there's no defined interfaces, then the type library follows the VBx rules.
If interfaces are specified, then type library generation follows tB(/VB.Net?) rules.

I'm sure there's added complexity I'm oversimplifying there though.

**Comment by bclothier** [comment]
But that's not what the issue is about. If you look at the OP, I imported a coclass from an external type library and then tried to implement it in my tB code. The resulting tB type library does _not_ indicate that it implements the original types from the original type library. The interface `IFoo` remains missing in the `Bar` class.

**Comment by mburns08109** [comment]
So, is this because it gets interpreted "the VBx way" (only)?

perhaps we need to be able to override the "default" way of reading the interface with an "Option AllInterfaces" sort of choice (as opposed to the "Option VbxInterfaces Only" default)? (not that source file Option keywords are what I really mean here, but something more in the "along those lines of thinking" way)
